lib/phoenix/tracker.ex

```diff
@@ -245,6 +245,33 @@ defmodule Phoenix.Tracker do
     |> Phoenix.Tracker.Shard.dirty_get_by_key(topic, key)
   end
 
+  @spec dirty_get_by_key(atom, term) :: [{topic, pid, meta :: map()}]
+  def dirty_get_by_key(tracker_name, key) do
+    0..(pool_size(tracker_name) - 1)
+    |> Task.async_stream(
```

Comment:
Might make sense to micro-benchmark this via. sequential fetch.
In some cases (especially if the number of shards is small), the penalty of starting and waiting for the task might outweigh the benefits of concurrent execution

Reply:
Interesting, but this one is more scalable w/ increasing pool size (if we do so in the future) right? Never benchmarked for comparison but since we are doing ets query, thought concurrent execution might be beneficial regardless since it's IO and that takes more time than the penalty you mentioned

Reply:
I am curious to know whether there would be any visible difference between using asynchronous and synchronous methods. For instance, when broadcasting messages to a large number of processes, it is more cost-effective to do it in a serial manner as opposed to using asynchronous methods. This method does not affect the tail latency or message delivery latency.

Reply:
Oh interesting, TIL (thought that some level of concurrency is used and not pure serialized sending of msgs). Thought spinning up for example 5 tasks to take )divide sending 1000 msgs would yield benefits (not necessarily scaling the tasks 1-1 but some concurrency helps

Reply:
To make it clear, it is more efficient to process messages sequentially rather than creating a separate task for each message. However, there may come a time when dividing the workload into several tasks, equivalent to the number of CPUs or a multiple of them, would be significantly advantageous.

Reply:
It all depends on number of recipients, so that's what I was curious of measuring to see how it shapes here.

Reply:
True true good point, agree with that - I guess if broadcasting is done completely sequentially, can probably do this sequentially for now to expose this API and unblock viewer list work and can explore benchmarking

Reply:
I set up a benchmark where I simulated 45k users and 600 lives (our peak numbers), with users equally split between the lives. Got some interesting results! 

### with one process querying the tracker:

```
Name            ips        average  deviation         median         99th %
async        575.12        1.74 ms    ±19.76%        1.59 ms        2.69 ms
sync         351.41        2.85 ms    ±11.25%        2.72 ms        4.00 ms

Comparison:
async        575.12
sync         351.41 - 1.64x slower +1.11 ms
```

### with a 100 processes

```
Name            ips        average  deviation         median         99th %
sync          28.49       35.10 ms    ±30.73%       33.85 ms       76.71 ms
async         23.86       41.92 ms    ±23.58%       39.48 ms       79.42 ms

Comparison:
sync          28.49
async         23.86 - 1.19x slower +6.82 ms
```

### with a 1000 processes

```
Name            ips        average  deviation         median         99th %
sync           3.03      329.99 ms    ±18.83%      321.60 ms      474.23 ms
async          2.39      418.39 ms    ±16.67%      408.22 ms      587.24 ms

Comparison:
sync           3.03
async          2.39 - 1.27x slower +88.40 ms
```

---

As soon as we enable any kind of parallelism, the async version loses its advantage. We will be querying the tracker from multiple processes at once, that's for sure, so the sync version wins here. I'll update the code.

Reply:
Oh that's really interesting to know, thanks for this benchmark!

Reply:
Out of curiosity, why do you think async ends up being slower? Is it because it's simulating as if it's even more processes hitting it so its better to do sync?

Reply:
My guess would be that it's the overhead from copying the results of querying the ETS tables across the process boundaries + the contention on the table. @studzien correctly noticed that we didn't have `read_concurrency` enabled on the tracker's ETS table, but even that didn't help.

Reply:
Ah got it good to know, thanks for benchmarking!